File: homework/hw2/python/queens_simulated_anneal.py
from math import exp, log
from time import time

from problem.queens_move import QueensMoveState
from problem.queens_swap import QueensSwapState
from algorithm.simulated_anneal import SimulatedAnneal

if __name__ == '__main__':
    n = 67
    log_n = log(n)
    start_temp_log = n / log(n)
    max_conflicts = n * (n - 1) >> 1
    value_estimator_swap = lambda state: max_conflicts - state.state.conflicts
    value_estimator_move = lambda state: max_conflicts - state.conflicts

    temperature_schedule_move = lambda time: exp(start_temp_log - time / (n << 4))
    temperature_schedule_swap = lambda time: exp(start_temp_log - time / n)

    t0 = time()

    # q = QueensSwapState(n)
    # sa = SimulatedAnneal(q)
    # sa.search(value_estimator_swap, temperature_schedule_swap, n << 2, max_conflicts, 1e-16)

    q = QueensMoveState(n)
    sa = SimulatedAnneal(q)
    sa.search(value_estimator_move, temperature_schedule_move, n << 2, max_conflicts, 1e-16)
    # terminate_temperature为常数1e-16 基本要达到16n*(n+16)才到停止探索，可忽略

    print(f"Total time: {time() - t0}")

File: homework/hw2/c++/queens_simulated_anneal.cpp
#include <cmath>

#include "problem/queens_move.hpp"
#include "problem/queens_swap.hpp"
#include "algorithm/simulated_anneal.hpp"

int n = 100;
int max_conflicts = n * (n-1) >> 1;

// 目标值：state.state().conflicts() == 0，即函数返回max_conflicts
double value_estimator_swap(const QueensSwapState& state){
    return max_conflicts - state.state().conflicts();
}

double value_estimator_move(const QueensMoveState& state){
    return max_conflicts - state.conflicts();
}

// 温度随时间变化的方式也可以尝试修改
double temperature_schedule_move(int time){
    static double log_n = log(n);
    static double start_temp_log = n/log_n;
    return exp(start_temp_log - double(time) / (n << 4));
}

double temperature_schedule_swap(int time){
    static double log_n = log(n);
    static double start_temp_log = n/log_n;
    return exp(start_temp_log - double(time) / n);
}

int main(){

    time_t t0 = time(nullptr);

    std::ios::sync_with_stdio(false);

    // 用QueensSwapState的问题建模（动作为交换两行的皇后）
    QueensSwapState q(n);
    SimulatedAnneal<QueensSwapState> sa(q);
    // 尝试4n次模拟退火，终态温度为10^-16
    sa.search(value_estimator_swap, temperature_schedule_swap, n << 2, max_conflicts, 1e-16); 
    
    // 用QueensMoveState的问题建模（动作为将某一行的皇后移动到某一列）
    //QueensMoveState q(n);
    //SimulatedAnneal<QueensMoveState> sa(q);
    //sa.search(value_estimator_move, temperature_schedule_move, n << 4, max_conflicts, 1e-18);

    std::cout << "Total time: " << time(nullptr) - t0 << std::endl;

    return 0;
}